### PR TITLE
8.0 [NEW] contract_invoice_merge_by_partner

### PR DESCRIPTION
**Pull request by carlosdauden**
This module merges same partner invoices generated by contracts

### CODE REVIEW

File: contract_invoice_merge_by_partner/README.rst [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+#. Go to *Invoicing > Customer Invoices* and show merged invoice
+
+.. image:: https://odoo-community.org/website/image/ir.attachment/5784_f2813bd/datas
+   :alt: Try me on Runbot
+   :target: https://runbot.odoo-community.org/runbot/95/8.0
+
+Known issues / Roadmap
+======================
+
+* If product supplier info min quantity is greater than procurement qty and we
```
Comment (pedrobaeza) [inline contract_invoice_merge_by_partner/README.rst]: Remove this

File: contract_invoice_merge_by_partner/README.rst [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+============
+
+To install this module you need *account_invoice_merge*, available in:
+
+* Install repository `OCA/account-invoicing <https://github.com/OCA/account-invoicing>`_.
+
+Usage
+=====
+
+#. Go to *Sales > Contracts* and create some contrats with same partner
```
Comment (pedrobaeza) [inline contract_invoice_merge_by_partner/README.rst]: s/contrats/contracts

File: contract_invoice_merge_by_partner/models/account_analytic_analysis.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+from openerp import api, models
+
+
+class PurchaseOrderLine(models.Model):
+    _inherit = 'account.analytic.account'
+
+    @api.multi
+    def _recurring_create_invoice(self, automatic=False):
+        invoice_obj = self.env['account.invoice']
+        invoices = invoice_obj.browse(
```
Comment (pedrobaeza) [inline contract_invoice_merge_by_partner/models/account_analytic_analysis.py]: Can this return another value different from the list of IDs. You should take it into account.

File: contract_invoice_merge_by_partner/models/account_analytic_analysis.py [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+            super(PurchaseOrderLine, self)._recurring_create_invoice(automatic)
+        )
+        res = []
+        unlink_list = []
+        for partner in invoices.mapped('partner_id'):
+            inv_to_merge = invoices.filtered(lambda x: x.partner_id == partner)
+            if len(inv_to_merge) > 1:
+                invoices_info = inv_to_merge.do_merge(
+                    keep_references=True, date_invoice=False)
```
Comment (pedrobaeza) [inline contract_invoice_merge_by_partner/models/account_analytic_analysis.py]: Don't keep references in this case

**Comment by pedrobaeza** [comment]
You should allow to select which partners want to merge their invoices through a flag in res.partner.

File: contract_invoice_merge_by_partner/models/account_analytic_analysis.py [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+                res.extend(invoices_info.keys())
+                for inv_ids_list in invoices_info.values():
+                    unlink_list.extend(inv_ids_list)
+            else:
+                res.append(inv_to_merge.id)
+        if unlink_list:
```
Comment (pedrobaeza) [inline contract_invoice_merge_by_partner/models/account_analytic_analysis.py]: I think you don't need this if you don't keep references.

File: contract_invoice_merge_by_partner/models/account_analytic_analysis.py (hunk 3 above)
Comment (carlosdauden) [inline contract_invoice_merge_by_partner/models/account_analytic_analysis.py]: I don't return invoices var.
If you are referring to super, it seems that always return a list or IDs https://github.com/odoo/odoo/blob/8.0/addons/account_analytic_analysis/account_analytic_analysis.py#L791

File: contract_invoice_merge_by_partner/models/account_analytic_analysis.py (hunk 5 above)
Comment (carlosdauden) [inline contract_invoice_merge_by_partner/models/account_analytic_analysis.py]: keep_references only keep/don't keep invoice name. https://github.com/OCA/account-invoicing/blob/8.0/account_invoice_merge/invoice.py#L130

Comment (pedrobaeza) [inline contract_invoice_merge_by_partner/models/account_analytic_analysis.py]: OK, good to know

**Comment by carlosdauden** [comment]
@pedrobaeza Flag in res.partner added to select which partners want to merge their invoices.

File: contract_invoice_merge_by_partner/README.rst (hunk 2 above)
Comment (elicoidal) [inline contract_invoice_merge_by_partner/README.rst]: remove 'some'

File: contract_invoice_merge_by_partner/models/account_analytic_analysis.py [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+class PurchaseOrderLine(models.Model):
+    _inherit = 'account.analytic.account'
+
+    @api.multi
+    def _recurring_create_invoice(self, automatic=False):
+        invoice_obj = self.env['account.invoice']
+        invoices = invoice_obj.browse(
+            super(PurchaseOrderLine, self)._recurring_create_invoice(automatic)
+        )
+        res = []
```
Comment (pedrobaeza) [inline contract_invoice_merge_by_partner/models/account_analytic_analysis.py:17]: As this is new API, you are receiving a recordset of invoices, and so, you must return another invoices recordset.

**Comment by pedrobaeza** [comment]
cc @Tecnativa

**Comment by rafaelbn** [comment]
Tested in http://3148909-33-7a3b4d.runbot2.odoo-community.org

I have used as in the readme but invoices are not merged:

File: contract_invoice_merge_by_partner/models/res_partner.py [hunk 7]
```diff
@@ -0,0 +1,11 @@
...
+# © 2016 Carlos Dauden <carlos.dauden@tecnativa.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from openerp import fields, models
+
+
+class ResPartner(models.Model):
+    _inherit = 'res.partner'
+
+    contract_invoice_merge = fields.Boolean()
```
Comment (dreispt) [inline contract_invoice_merge_by_partner/models/res_partner.py]: Can you provide a string text to the field?

File: contract_invoice_merge_by_partner/models/account_analytic_analysis.py [hunk 8]
```diff
@@ -0,0 +1,30 @@
+# -*- coding: utf-8 -*-
+# © 2016 Carlos Dauden <carlos.dauden@tecnativa.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from openerp import api, models
+
+
+class PurchaseOrderLine(models.Model):
+    _inherit = 'account.analytic.account'
```
Comment (dreispt) [inline contract_invoice_merge_by_partner/models/account_analytic_analysis.py:10]: Is the class name correct? Shuldn't it be `AnalyticAccount` ?

**Comment by dreispt** [comment]
Made PR https://github.com/Tecnativa/contract/pull/1

**Comment by rafaelbn** [comment]
Could you review this PR @carlosdauden  please?

Check https://github.com/OCA/contract/pull/33#issuecomment-228784121

cc @sergio-teruel @pedrobaeza

**Comment by rafaelbn** [comment]
After https://github.com/Tecnativa/contract/pull/1 travis fails

**Comment by pedrobaeza** [comment]
@rafaelbn @carlosdauden, PR fixed. Please review again.

**Comment by carlosdauden** [comment]
:+1: Thanks for your improvements

**Comment by rafaelbn** [comment]
I followed againt steps of section `Usage` in README and I get same proble as in https://github.com/OCA/contract/pull/33#issuecomment-228784121

I don't get https://github.com/OCA/contract/pull/33/files#diff-e6b1771de6d8dbe760648e77e93caf3cR25

Could you please review it @carlosdauden ?

**Comment by rafaelbn** [comment]
@carlosdauden @pedrobaeza how is the state of this PR? Could you review https://github.com/OCA/contract/pull/33#issuecomment-234422449 please?

**Comment by pedrobaeza** [comment]
@rafaelbn, the module is now perfectly working. You should be doing something wrong. Maybe not the same next invoicing date?

**Comment by rafaelbn** [comment]
It doesn't work
- Go to _Sales > Contracts_ and create some contracts with same partner (I created 3)
- Go to _Settings > Automation > Scheduled Actions_
- Select _Generate Recurring Invoices from Contracts_
- Set previous time that now in _Next Execution Date_
- Go to _Invoicing > Customer Invoices_ and 3 contracts -> 3 invoces

**Comment by pedrobaeza** [comment]
You're again forgetting to check the mark in the partner.

File: contract_invoice_merge_by_partner/README.rst [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+Installation
+============
+
+To install this module you need *account_invoice_merge*, available in:
+
+* Install repository `OCA/account-invoicing <https://github.com/OCA/account-invoicing>`_.
+
+Usage
+=====
+
```
Comment (rafaelbn) [inline contract_invoice_merge_by_partner/README.rst]: Go to Sales > Customer > Choose a partner and in the tab "Sales & Purchases" activate check " Merge contracts invoices "

File: contract_invoice_merge_by_partner/README.rst [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+#. Go to *Sales > Contracts* and create some contracts with same partner
```
Comment (rafaelbn) [inline contract_invoice_merge_by_partner/README.rst]: s/with same partner/with same partner you checked the checkbox " Merge contracts invoices "

**Comment by rafaelbn** [comment]
OK. Tested again and now is :+1:

**Pull request merged**